Dev container Templates & distribution (#98)
devcontainers/spec: proposals/devcontainer-templates.md

```
+-- template
|    +-- devcontainer-template.json
|    +-- .devcontainer
|        +-- devcontainer.json 
|        +-- Dockerfile (optional)
|        +-- docker-compose.yml (optional)
|        +-- (other files)
|    +-- (other files)
```

## devcontainer-template.json properties

The `devcontainer-template.json` file defines information about the Template to be used by any [supporting tools](#template-supporting-tools-and-services).

The properties of the file are as follows:

| Property | Type | Description |
| :--- | :--- | :--- |
| `id` | string | ID of the Template. The `id` should be unique in the context of the repository/published package where the Template exists and must match the name of the directory where the `devcontainer-template.json` resides. |
| `version` | string | The semantic version of the Template. |
| `name` | string | Name of the Template. |
| `description` | string | Description of the Template. |
| `documentationURL` | string | Url that points to the documentation of the Template. |
| `licenseURL` | string | Url that points to the license of the Template. |

> | `type` | string | Type of the dev container (image, dockerfile, dockerCompose) created by the Template. |

Reviewer: How is this used?

Author: Currently, the `definitions` have this `type` property (`singleContainer` and `dockerCompose`) which is used by the UI to determine how to start or stop the container and configure few other things. Hence, added it here. 

Also, as per today's discussion, this will also be useful for listing Templates and people can sort with the `type`

Reviewer 2: Perhaps the tooling could add some of these attributes.  I imagine from some `devcontainer template package` command, it would not be too difficult to programatically add several of these attributes (and ensures consistency/doesn't require action from the author).

Author: Just wondering what if the tooling mistags some times ?

Example - lets say a Template contains all of three files (.devcontainer.json, dockerfile and dockercompose), the Template author writes code in such a way that the dev container solely depends on an image and has only added boiler plate code for `dockerfile` and `dockerCompose` or samples of how to configure using them. If the tooling simply determines the value based on the file presence of some other attributes, I wonder it might mis configure sometimes 🤔

Author: Makes sense. Removed `type`

Reviewer: Sorry, I wasn't clear: I suggest to keep `type` and we can add tooling support later (whenever we get to it). :)

> | `image` | string | Name of an image in a container registry ([DockerHub](https://hub.docker.com/), [GitHub Container Registry](https://docs.github.com/en/packages/learn-github-packages/introduction-to-github-packages), [Azure Container Registry](https://azure.microsoft.com/en-us/products/container-registry/)) that the [Template supporting services / tools](#template-supporting-tools-and-services) should use to populate list of image variants. |

Reviewer: Is this for the UI? Can `options` be used for this instead of the additional `image` property?

Author: Removed as per the discussion and added to `options` instead.

> | `features` | object | List of [Feature id](devcontainer-features.md/#referencing-a-feature) that the [Template supporting services / tools](#template-supporting-tools-and-services) should use to populate different configuration options. |

Reviewer: Is this for showing a list to the user to pick from? Would it include the major version?

Author: Removed as per the discussion and added to options instead.

| `options` | object | A map of options that will be passed as `build.args` when building a dockerfile. |
| `platforms` | object | Languages and platforms supported by the Template. |
| `keywords` | array | List of strings relevant to a user that would search for this Template. |

### The `options` property

The options property contains a map of option IDs and their related configuration settings. The ID becomes the name of the build arguments in all caps and snake case which is passed while building a dockerfile. See [option resolution](#option-resolution) for more details. For example:

```json
{
  "options": {
    "optionIdGoesHere": {
      "type": "string",
      "description": "Description of the option",
      "proposals": ["value1", "value2"],
      "default": "value1"
    }
  }
}
```

| Property | Type | Description |
| :--- | :--- | :--- |
| `optionId` | string | ID of the option that is converted into an all-caps and snake case build argument with the selected value in it. |

> | `optionId.type` | string | Type of the option. Valid types are currently: `boolean`, `string` |

Reviewer: Build args are always strings (similar to env variables).

Template metadata is captured by a `devcontainer-template.json` file in the root folder of the Template.

## Folder structure 

A Template is a self contained entity in a folder with at least a `devcontainer-template.json` and [`devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson).  Additional files are permitted and are packaged along side the required files.

```
+-- template
|    +-- devcontainer-template.json
|    +-- .devcontainer.json
|    +-- (other files)
```

> ## devcontainer-template.json properties

Author: @Chuxel  For the `publisher` info which is needed by the UI -  I believe we will have a similar index for Templates as we have for [Features](https://github.com/devcontainers/devcontainers.github.io/blob/gh-pages/_data/collection-index.yml#L2) and this index will be used by the tooling to display list of Templates. We can then make use of `maintainer` property to display to `publisher/maintainer` name wherever needed.  

 Adding `publisher` prop to `devcontainer-template.json` feels a little unsafe as any author can claim to be someone else and we currently don't have identity checking mechanism in place. At least, with the search index, we will manually look at the PR created by the author (to add his Templates to crawler) and confirm if the maintainer prop matches with  other details.

Author: Agreed, added `publisher` 👍

# Dev Container Templates reference

> Development container "Templates" are source files packaged together that encode configuration for a complete development environment. A Template can be used in a new or existing project and a [supporting tool](https://containers.dev/supporting) will use the configuration from the template to build a development container.

Reviewer 3: ```suggestion
Development container "Templates" are source files packaged together that encode configuration for a complete development environment. A Template can be used in a new or existing project, and a [supporting tool](https://containers.dev/supporting) will use the configuration from the Template to build a development container.
```

> The configuration is placed in a [`.devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson) which can also reference other files within the Template. Alternatively, `.devcontainer/devcontainer.json` can also be used if the container needs to reference other files, such as a `Dockerfile` or `docker-compose.yml`. A template can also provide additional source files (eg: boilerplate code or a [lifecycle script](https://containers.dev/implementors/json_reference/#lifecycle-scripts).

Reviewer 3: ```suggestion
The configuration is placed in a [`.devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson) which can also reference other files within the Template. Alternatively, `.devcontainer/devcontainer.json` can also be used if the container needs to reference other files, such as a `Dockerfile` or `docker-compose.yml`. A Template can also provide additional source files (eg: boilerplate code or a [lifecycle script](https://containers.dev/implementors/json_reference/#lifecycle-scripts).
```

| `optionId.default` | string | Default value for the option. |
| `optionId.replaceIn` | array | List of file paths which the supporting tool should use to perform the string replacement of `optionId` with the selected value. The provided path is always relative to the sub-directory of the Template. |

> `Note`: The `options` must be unique for every `devcontainer-template.json`

### Referencing a Template 

The `id` format (`<oci-registry>/<namespace>/<template>[:latest]`) dictates how a [supporting tool](../docs/specs/supporting-tools.md#supporting-tools-and-services) will locate and download a given Template from an OCI registry (For example - `ghcr.io/user/repo/go:latest`). The registry must implement the [OCI Artifact Distribution Specification](https://github.com/opencontainers/distribution-spec). Some implementors can be [found here](https://oras.land/implementors/). 

## Versioning

Each Template is published with only the `latest` tag. Tooling that handles releasing Templates will republish the `latest` tag every time a new release is created.

## Release

> _For information on distribution Templates, see [devcontainer-templates-distribution.md](./devcontainer-templates-distribution.md)._

Reviewer 3: ```suggestion
_For information on distributing Templates, see [devcontainer-templates-distribution.md](./devcontainer-templates-distribution.md)._
```

### Option Resolution

> A templates's `options` is used by a supporting tool to prompt for different configuration options. A supporting tool will parse the `options` object provided by the user. If a value is selected for a Template, it will be replaced in the provided `replaceIn` files.

Reviewer 3: ```suggestion
A Template's `options` property is used by a supporting tool to prompt for different configuration options. A supporting tool will parse the `options` object provided by the user. If a value is selected for a Template, it will be replaced in the provided `replaceIn` files.
```

### Option resolution example

> Consider a `java` template with the following folder structure:

Reviewer 3: ```suggestion
Consider a `java` Template with the following folder structure:
```

        "description": "Specify version of java.",
        "proposals": [
          "17-bullseye",
          "17-buster",
          "11-bullseye",
          "11-buster",
          "17",
          "11"
        ],
			  "default": "17-bullseye",
        "replaceIn": [".devcontainer.json"]
    },
    "nodeVersion": {
        "type": "string", 
        "proposals": [
          "latest",
          "16",
          "14",
          "10",
          "none"
        ],
        "default": "latest",
        "description": "Specify version of node, or 'none' to skip node installation.",
        "replaceIn": [".devcontainer.json"]
    },
    "installMaven": {
        "type": "boolean", 
        "description": "Install Maven, a management tool for Java.",
        "default": "false",
        "replaceIn": [".devcontainer.json"]
    },
}
```

and it has the following `.devcontainer.json` file:

```json
{
	"name": "Java",
	"image": "mcr.microsoft.com/devcontainers/java:0-${templateOption:imageVariant}",
	"features": {
		"ghcr.io/devcontainers/features/node:1": {
			"version": "${templateOption:nodeVersion}",
      "installMaven": "${templateOption:installMaven}"
		}
	},
//	...
}
```

> A user tries to add the `java` Template to their project using the [supporting tools](../docs/specs/supporting-tools.md#supporting-tools-and-services) and selects `17-bullseye` when prompted for `"Specify version of Go."` and uses the `default` values for `"Specify version of node, or 'none' to skip node installation."` and `"Install Maven, a management tool for Java."`.

Reviewer 3: ```suggestion
A user tries to add the `java` Template to their project using the [supporting tools](../docs/specs/supporting-tools.md#supporting-tools-and-services) and selects `17-bullseye` when prompted for `"Specify version of Go"` and the `default` values when prompted for `"Specify version of node, or 'none' to skip node installation"` and `"Install Maven, a management tool for Java"`.
```

# Dev Container Templates reference

> Development container "Templates" are self-contained, shareable units of development container configuration. The name comes from the idea that adding a Template to your project helps it get up and running with a containerized environment. The Templates describe the appropriate container image, [Features](https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md), runtime arguments for starting the container, and VS Code extensions that should be installed. Each provides a container configuration file ([`.devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson)) and other needed files that you can drop into any existing folder as a starting point for containerizing your project.

Reviewer 2: ```suggestion
Development container "Templates" are source files packaged together that encode configuration for a complete development environment. A Template can be used in a new or existing project.  A [supporting tool](https://containers.dev/supporting) will use the configuration from the template to build a development container.

Configuration is placed in a [`.devcontainer/devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson). The `devcontainer.json` can reference other files within the template, such as a `Dockerfile` or `docker-compose.yml`. 

A template can also provide additional example source files (eg: boilerplate code or a [lifecycle script](https://containers.dev/implementors/json_reference/#lifecycle-scripts).

```

Author: Thanks for the edit. looks good, added with little tweaks

Template metadata is captured by a `devcontainer-template.json` file in the root folder of the Template.

## Folder structure 

> A Template is a self contained entity in a folder with at least a `devcontainer-template.json` and [`devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson).  Additional files are permitted and are packaged along side the required files.

Reviewer 2: ```suggestion
A single Template is a folder with at least a `devcontainer-template.json` and [`devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson).  Additional files are permitted and are packaged along side the required files.
```

# Dev Container Templates reference

Development container "Templates" are source files packaged together that encode configuration for a complete development environment. A Template can be used in a new or existing project and a [supporting tool](https://containers.dev/supporting) will use the configuration from the template to build a development container.

> Configuration is placed in a [`.devcontainer.json`](/docs/specs/devcontainer-reference.md#devcontainerjson). Alternatively, `.devcontainer/devcontainer.json` can also be used if the container needs to reference other files within the template, such as a `Dockerfile` or `docker-compose.yml`. A template can also provide additional source files (eg: boilerplate code or a [lifecycle script](https://containers.dev/implementors/json_reference/#lifecycle-scripts).

Reviewer: Can a `.devcontainer.json` also reference other files? (I thought so, but the wording seems to suggest otherwise.)

Prepend "The" to "configuration"?

Author: Good catch, updated wordings.

      "proposals": ["value1", "value2"],
      "default": "value1",
      "replaceIn": [".devcontainer.json"],
    }
  }
}
```

| Property | Type | Description |
| :--- | :--- | :--- |
| `optionId` | string | ID of the option used by the supporting tools to replace the selected value in the specified `replaceIn` files. |
| `optionId.type` | string | Type of the option. Valid types are currently: `boolean`, `string` |
| `optionId.description` | string | Description for the option. |
| `optionId.proposals` | array | A list of suggested string values. Free-form values **are** allowed. Omit when using `optionId.enum`. |
| `optionId.enum` | array | A strict list of allowed string values. Free-form values are **not** allowed. Omit when using `optionId.proposals`. |
| `optionId.default` | string | Default value for the option. |
| `optionId.replaceIn` | array | List of file paths which the supporting tool should use to perform the string replacement of `optionId` with the selected value. The provided path is always relative to the sub-directory of the Template. |

> `Note`: The `options` must be unique for every `devcontainer-template.json`

> ### Template Supporting Tools and Services

Reviewer 4: As with the next section, I'm not sure we need to reference this as a part of the proposal as much as it should be in docs for VS Code and Codespaces.  We could, however, add this into the spec's "implementing tools" doc.  

You could refactor this to be "example usage" if you think it's important, but it might be better just to omit and add the implementing tools part when we move it into containers.dev.

Author: Makes sense, removed. 👍

Outlines tools and services that currently support adding Templates to a project.

- [Visual Studio Code Remote - Containers](../docs/specs/supporting-tools.md/#visual-studio-code-remote---containers)
- [GitHub Codespaces](../docs/specs/supporting-tools.md/#github-codespaces)

## Adding a Template to a project or codespace

>   

Reviewer 4: Since this section is tool / service specific, I'm not sure it belongs in the spec as much as it does documentation for VS Code and Codespaces. It could be added to tool specific steps.  A CLI-based example would be better if there's a need to have something here.

Author: Makes sense, removed!

```json
// ...
"options": {
    "imageVariant": {
        "type": "string",
        "description": "Specify version of java.",
        "proposals": [
          "17-bullseye",
          "17-buster",
          "11-bullseye",
          "11-buster",
          "17",
          "11"
        ],
			  "default": "17-bullseye",
        "replaceIn": [".devcontainer.json"]
    },
    "nodeVersion": {
        "type": "string", 
        "proposals": [
          "latest",
          "16",
          "14",
          "10",
          "none"
        ],
        "default": "latest",
        "description": "Specify version of node, or 'none' to skip node installation.",
        "replaceIn": [".devcontainer.json"]
    },
    "installMaven": {
        "type": "boolean", 
        "description": "Install Maven, a management tool for Java.",
        "default": "false",
        "replaceIn": [".devcontainer.json"]
    },
}
```

and it has the following `.devcontainer.json` file:

```json
{
	"name": "Java",
	"image": "mcr.microsoft.com/devcontainers/java:0-{{imageVariant}}",
	"features": {
		"ghcr.io/devcontainers/features/node:1": {

> 			"version": "{{nodeVersion}}",

Reviewer: We are using `${localEnv:<env var name>}` and similar variables at runtime. Should we use `${templateOption:<option id>}` (e.g., `${templateOption:nodeVersion}` here) for consistency?

Author: Nice, updated!
